manager/src/main/java/org/openremote/manager/asset/ForecastService.java

```diff
@@ -0,0 +1,801 @@
+            while (true) {
+                synchronized (ForecastTaskManager.this) {
+                    if (scheduledFuture == null) {
+                        return true;
+                    } else if (scheduledFuture != null && scheduledFuture.getDelay(TimeUnit.MILLISECONDS) > DELAY_MIN_TO_CANCEL_SAFELY) {
+                        scheduledFuture.cancel(false);
+                        scheduledFuture = null;
+                        return true;
+                    } else {
+                        if (timerService.getCurrentTimeMillis() - start > timeout) {
+                            scheduledFuture.cancel(true);
+                            scheduledFuture = null;
+                            return false;
+                        }
+                    }
+                }
+                try {
+                    Thread.currentThread().sleep(300);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    return false;
+                }
+            }
+        }
+
+        private synchronized void start(long now) {
+            start(now, false);
+        }
+
+        private synchronized void start(long now, boolean isServerRestart) {
+            if (scheduledFuture == null) {
+                addForecastTimestamps(now, isServerRestart);
+                updateNextForecastCalculationMap();
+                scheduleForecastCalculation(now, Optional.empty());
+            }
+        }
+
+        private void calculateForecasts() {
+            final long now = timerService.getCurrentTimeMillis();
+            Map<AttributeRef, Pair<ForecastConfiguration, List<Long>>> configAndTimestampMap = new HashMap<>();
+
+            try {
+                synchronized (ForecastTaskManager.this) {
+                    if (Thread.currentThread().isInterrupted()) {
+                        return;
+                    }
+                    purgeForecastTimestamps(now);
+                    addForecastTimestamps(now, false);
+                    purgeForecastTimestamps(now);
+
+                    nextForecastCalculationMap.forEach((attributeRef, nextForecastCalculationTimestamp) -> {
+                        if (nextForecastCalculationTimestamp <= now && forecastTimestampMap.containsKey(attributeRef) && configMap.containsKey(attributeRef)) {
+                            configAndTimestampMap.put(
+                                attributeRef,
+                                new Pair<>(configMap.get(attributeRef), forecastTimestampMap.get(attributeRef))
+                            );
+                        }
+                    });
+                }
+
+                configAndTimestampMap.forEach((attributeRef, configAndTimestamps) -> {
+                    Optional<Attribute<?>> attribute = Optional.empty();
+                    Asset<?> asset = withLockReturning(getClass().getSimpleName() + "::calculateForecasts", () ->
```

Comment:
I thought that using `AttributeRef` is the most natural way to reference attributes that have a forecast configuration. I don't think that I can replace `AttributeRef` with the `Attribute` class as key in the config map because the `equals()` implementation of Attribute is not suitable for that. I think the easiest way to make it more efficient is to store the `Attribute` instance in an additional map so that I do not have to retrieve it by means of the asset storage service.